### PR TITLE
3.x: Change Flowable.groupBy to signal MBE instead of possibly hanging

### LINKED ISSUE #6641: 3.x: Change the behavior of Flowable.groupBy to signal MBE if no main requests
Currently, the `Flowable.groupBy` queues up groups until the downstream requests more groups. Unfortunately, this can lead to hangs because if there are more groups than requested, active groups may starve out as they don't trigger replenishments or those replenishments result in more groups to be created. Example:

```java
Flowable.range(1, 1000)
.groupBy(v -> v)
.flatMap(v -> v, 16)
.test()
.assertValueCount(1000)
.assertComplete();
```
This case works if `flatMap`'s concurrency level is greater or equal to the expected group count.

I propose changing the `Flowable.groupBy` operator to signal `MissingBackpressureException` if the direct downstream is not ready to receive more groups.

### PR DESCRIPTION
**Pull request by akarnokd**
This PR changes the backpressure behavior of `Flowable.groupBy` to signal `MissingBackpressureException` instead of silently hanging if the produced groups are not ready to be accepted by the downstream. 

This can happen if one `flatMap`s a `groupBy` but there are more groups produced than the concurrency level of `flatMap`. Since replenishment is tied to item consumption from the groups, not consuming them can result in none of the groups receiving any further items and the whole operator hangs.

The following changes have been applied:

- Removed the queue from the main operator since it will now try to emit directly and not buffer groups.
- The main `Flowable`, lacking a queue, no longer supports operator fusion. Tests checking this property have been removed as well.
- When a group is drained, consumed items are replenished in batch if possible. Detecting a cancellation will also trigger a replenishment.
- When a group is pulled (fusion mode), now all `pull`, `isEmpty` and `clear` will trigger replenishment so that other groups can make progress too.
- Unit tests have been modified to have large enough bufferSize/prefetch amounts to allow them to pass.

Fixes #6641

### CODE REVIEW

**Review by vanniktech — APPROVED** [review]
Never used that operator and I remember @davidmoten had lots of use cases so let's wait for his approval?

**Comment by akarnokd** [comment]
Sure.

**Comment by akarnokd** [comment]
Let's merge this into the release and get feedback from the field.

**Pull request merged**
